src/index.ts

```diff
@@ -74,9 +75,15 @@ export function parseForESLint(
             ecmaVersion: options.ecmaVersion || DEFAULT_ECMA_VERSION,
         }
         const skipParsingScript = options.parser === false
-        const tokenizer = new HTMLTokenizer(code, optionsForTemplate)
-        const rootAST = new HTMLParser(tokenizer, optionsForTemplate).parse()
-
+        let tokenizer: HTMLTokenizer | PugTokenizer
+        let rootAST: AST.VDocumentFragment
+        if (/^\<template\s+lang\=["']pug["']\>/i.test(code)) {
+            tokenizer = new PugTokenizer(code, optionsForTemplate)
+            rootAST = new PugParser(tokenizer, optionsForTemplate).parse()
+        } else {
+            tokenizer = new HTMLTokenizer(code, optionsForTemplate)
+            rootAST = new HTMLParser(tokenizer as HTMLTokenizer, optionsForTemplate).parse()
+        }
```

Comment:
My current draft tries to aim for compatibility to the current source for HTML
So I created a tokenizer and a parser for pug
These can later be extended and formed to their needs

Reply:
I'm not familiar with pug, but I think we need to parse the text inside the `<template>` element as a pug, after extracting the `<template>` element from the result of parsing with HTMLParser.

Reply:
The pug parser (`pug-lexer`) can handle the `<template>` block totally on it's own
The `<template lang="pug">` results as a `TextHtmlToken`
https://github.com/pugjs/pug/blob/6cca8f76befe7285eead0416ba4e2dbba825ab65/packages/pug-lexer/index.d.ts#L174
The part between the opening and closing of `TextHtmlToken` are then just normal pug tokens.

So maybe we just need partially the `HTMLTokenizer` and `HTMLParser` but only for the non-template blocks 🤔 
See https://github.com/vuejs/vue-eslint-parser/pull/123#issuecomment-903923916

Reply:
The block elements in the `.vue` file are different from HTML, so I think it probably won't work well with the pug parser.